rfcs/0009-field-arguments.md

```diff
@@ -0,0 +1,57 @@
+# Field Arguments
+
+The ability to provide arguments to fields broadens the use-cases that NDC can support.
+
+This can be seen as generalising `fields` to `functions` or, simply as adding context to fields.
+
+## Proposal
+
+This proposes an update to the NDC-Spec to allow arguments to fields.
+
+The motivation for this change is to generalise the API surface to allow more expressive queries - This should be very similar to collection arguments in practice but can apply to any field (especially nested), not just top-level collections.
+
+Some examples of why this might be useful:
+
+* JSON Operations: For a JSON field in a PG table, several JSON functions could be exposed as fields but they require arguments to be useful. Such as [#>](https://www.postgresql.org/docs/9.3/functions-json.html)
+* Vector Operations for LLMs etc.
+* Advanced Geo Operations
+* GraphQL federation: Forwarding Hasura GQL schemas over NDC will require this change if we want to expose root fields as commands instead of collections.
+
+The schema is extended with:
+
+```rust
+pub struct ObjectField {
+  ...
+  pub arguments: BTreeMap<String, ArgumentInfo>,
+}
+```
+
+and queries are extended with:
+
+```rust
+pub enum Field {
+  Column {
+    ...
+    arguments: Option<BTreeMap<String, Option<Argument>>>,
```

Comment:
Could we implement this like so instead:
```rust
pub enum Field {
  Column {
    ...
    #[serde(skip_serializing_if = "BTreeMap::is_empty", default)]
    arguments: BTreeMap<String, Argument>,
  }
}
```
This maintains the same semantics when serialized while being a nicer API when de-serialized.

Also removed the inner option per the conversation elsewhere about arguments needing to specify null-ability rather than being optional.

Reply:
7036c49..ec562f5